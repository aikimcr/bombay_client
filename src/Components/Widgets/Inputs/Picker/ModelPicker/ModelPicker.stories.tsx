import React from 'react';
import type { Meta, StoryObj } from '@storybook/react-webpack5';
import {
  ContextChanger,
  TestCollectionHandlers,
  TestCollectionOne,
} from '../../../../../testHelpers';

import { ModelPicker } from './ModelPicker';

const meta = {
  component: ModelPicker,
  tags: ['autodocs'],
} satisfies Meta<typeof ModelPicker>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  parameters: {
    msw: {
      handlers: TestCollectionHandlers,
    },
  },
  args: {
    id: 'storybook-picker',
    initialCollection: new TestCollectionOne({}),
  },
  render: (args) => (
    <ContextChanger initialLoggedIn={true}>
      <ModelPicker {...args} />
    </ContextChanger>
  ),
};
