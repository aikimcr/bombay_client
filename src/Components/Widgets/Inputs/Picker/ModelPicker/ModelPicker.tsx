import React, { useState } from 'react';
import { CollectionBase } from '../../../../../Model/CollectionBase';
import { ModelPickerList } from '../ModelPickerList';
import { PickerButton } from '../PickerButton';

interface ModelPickerProps {
  id: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  initialCollection: CollectionBase<any, any>;
}

export const ModelPicker: React.FC<ModelPickerProps> = ({
  id,
  initialCollection,
}): React.ReactElement => {
  const [openState, setOpenState] = useState(false);

  const openStateToggle = () => {
    setOpenState((oldState) => !oldState);
  };

  return (
    <div data-testid="model-picker" id={id}>
      <PickerButton
        id={`${id}-button`}
        buttonText={'dummy'}
        openState={openState}
        openStateToggle={openStateToggle}
        popoverTarget={`${id}-popover`}
      />
      <ModelPickerList
        id={`${id}-popover`}
        initialCollection={initialCollection}
      />
    </div>
  );
};
