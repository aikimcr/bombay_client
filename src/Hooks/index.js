export * from './useLoginTracking';
export * from './useModelCollection';
