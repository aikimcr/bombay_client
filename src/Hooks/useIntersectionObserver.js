import { useEffect, useRef } from "react";

export function useIntersectionObserver(topRef, callback, options = {}) {
  const observer = useRef(null);

  useEffect(() => {
    if (topRef.current == null) return;

    if (observer.current == null) {
      observer.current = new IntersectionObserver(callback, {
        ...options,
        root: topRef.current.parentElement,
      });
    }

    return () => {
      observer.current.disconnect();
    };
  }, [callback, options, topRef]);

  return observer;
}

export default useIntersectionObserver;
