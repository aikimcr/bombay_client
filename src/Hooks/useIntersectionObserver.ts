import { useEffect, useRef, RefObject } from 'react';

export interface IntersectionObserverOptions {
  root?: Element | null;
  rootMargin?: string;
  threshold?: number | number[];
}

export interface UseIntersectionObserverReturn {
  current: IntersectionObserver | null;
}

export function useIntersectionObserver(
  topRef: RefObject<Element>,
  callback: IntersectionObserverCallback,
  options: IntersectionObserverOptions = {},
): UseIntersectionObserverReturn {
  const observer = useRef<IntersectionObserver | null>(null);

  useEffect(() => {
    if (topRef.current == null) {
      return;
    }

    if (observer.current == null) {
      observer.current = new IntersectionObserver(callback, {
        ...options,
        root: topRef.current.parentElement,
      });
    }

    return () => {
      if (observer.current) {
        observer.current.disconnect();
      }
    };
  }, [callback, options, topRef]);

  return observer;
}

export default useIntersectionObserver;
