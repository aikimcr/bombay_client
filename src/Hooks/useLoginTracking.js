import { useCallback, useEffect, useRef, useState } from "react";

import { loginStatus, refreshToken } from "../Network/Login";
// Allow external forcing of the state
let setLoggedIn;
export function forceLoginState(newState) {
  if (setLoggedIn) return setLoggedIn(newState);
}

function toMilliseconds(minutes) {
  return minutes * 60 * 1000;
}

export function useLoginTracking() {
  const [loginState, setLoginState] = useState(false);
  const [activitySeen, setActivitySeen] = useState(true);

  const activityCheckMinutes = 30;
  const loginCheckMinutes = 10;

  let activityHandle = useRef(false);

  setLoggedIn = setLoginState; // Allow external forcing of the state.

  const checkLoginState = useCallback(async function () {
    const result = await loginStatus(activityCheckMinutes);
    setLoginState(result);
    return result;
  }, []);

  const refreshState = useCallback(
    async function () {
      if (!loginState) return false;

      const result = await refreshToken();
      setLoginState(result != null);
      return result != null;
    },
    [loginState],
  );

  const checkActivity = useCallback(function (evt) {
    // console.debug(`${new Date().toLocaleString()}: Check Activity (${evt}) (Activity: ${activitySeen}, loginState: ${loginState})`);

    clearTimeout(activityHandle.current);

    if (evt != null) setActivitySeen(true);

    activityHandle.current = setTimeout(async () => {
      // console.debug(`${new Date().toLocaleString()}: Clear Activity (Activity: ${activitySeen}, loginState: ${loginState})`);

      setActivitySeen(false);
    }, toMilliseconds(activityCheckMinutes));
  }, []);

  const setListeners = useCallback(() => {
    document.addEventListener("keydown", checkActivity);
    document.addEventListener("mousedown", checkActivity);
    document.addEventListener("mousemove", checkActivity);
  }, [checkActivity]);

  const removeListeners = useCallback(() => {
    document.removeEventListener("mousemove", checkActivity);
    document.removeEventListener("mousedown", checkActivity);
    document.removeEventListener("keydown", checkActivity);
  }, [checkActivity]);

  const checkLogin = useCallback(
    async function () {
      // console.debug(`${new Date().toLocaleString()}: Check login (Activity: ${activitySeen}, loginState: ${loginState})`);

      if (activitySeen) {
        await refreshState();
        return;
      }

      return await checkLoginState();
    },
    [activitySeen, refreshState, checkLoginState],
  );

  useEffect(() => {
    // console.debug(`${new Date().toLocaleString()}: Change (Activity: ${activitySeen}, loginState: ${loginState})`);
    let intervalHandle = false;

    if (loginState) {
      intervalHandle = setInterval(
        checkLogin,
        toMilliseconds(loginCheckMinutes),
      );
      checkActivity();
      setListeners();
    }

    return () => {
      removeListeners();
      clearTimeout(activityHandle.current);
      clearInterval(intervalHandle);
    };
  }, [
    activitySeen,
    loginState,
    activityHandle,
    checkActivity,
    checkLogin,
    removeListeners,
    setListeners,
  ]);

  useEffect(() => {
    checkLoginState();
  }, [checkLoginState]);

  return [loginState, setLoginState];
}

export default useLoginTracking;
