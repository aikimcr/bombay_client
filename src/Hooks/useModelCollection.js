import { useState, useEffect, useContext, useRef, useCallback } from 'react';

import BombayLoginContext from '../Context/BombayLoginContext';
import useIntersectionObserver from './useIntersectionObserver';

export const useModelCollection = ({ CollectionClass, topRef }) => {
  const { loggedIn } = useContext(BombayLoginContext);
  const [shouldPage, setShouldPage] = useState(false);
  const [loading, setLoading] = useState(loggedIn);

  const [collection, setCollection] = useState(null);
  const observer = useIntersectionObserver(topRef, (entries, observer) => {
    entries.forEach((entry) => {
      if (entry.isIntersecting) {
        setShouldPage(true);
        observer.unobserve(entry.target);
      }

      // I haven't found a case where this is actually needed, but I haven't ruled it out.
      // if (entry.isVisible) {
      // }
    });
  });

  const refreshCollection = async () => {
    // console.count(`refresh, ${loggedIn}`);
    if (!loggedIn) {
      setCollection(null);
      return;
    }

    setLoading(true);

    const newCollection = new CollectionClass();
    await newCollection.ready();
    // console.count(`newCollection ready: ${newCollection}`);

    setLoading(false);
    setCollection(newCollection);
  };

  useEffect(() => {
    // console.count(`mount hook, ${collection}, ${loading}`);
    if (collection == null) {
      refreshCollection();
    }
  }, []);

  useEffect(() => {
    // console.count(`shouldPage: ${shouldPage}, loading: ${loading}`);
    if (!shouldPage) {
      return;
    }
    if (loading) {
      return;
    }

    setShouldPage(false);

    if (collection.hasNextPage()) {
      setLoading(true);

      collection.fetchNextPage().then(() => {
        setLoading(false);
      });
    }
  }, [shouldPage, loading]);

  useEffect(() => {
    // console.count(`loading: ${loading}, observer: ${observer}, topRef: ${topRef}`);
    if (loading) {
      return;
    }

    if (collection && collection.hasNextPage()) {
      const myElement = topRef.current?.querySelector('li:last-child');

      if (myElement) {
        observer.current.observe(myElement);
      }
    }
  }, [loading, observer, topRef]);

  useEffect(() => {
    // console.count(`loggedIn change, ${loggedIn}, ${collection} ${loading}`);
    if (loading) {
      return;
    }

    if (loggedIn && collection == null) {
      refreshCollection();
      return;
    }

    if (!loggedIn) {
      setCollection(null);
    }
  }, [loggedIn]);

  // console.log('return collection', collection);
  return [collection, refreshCollection];
};

export default useModelCollection;
