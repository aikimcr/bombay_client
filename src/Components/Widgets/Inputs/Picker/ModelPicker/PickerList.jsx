import React, { createRef, useContext, useEffect, useState } from 'react';
import PropTypes from 'prop-types';

import BombayLoginContext from '../../../../../Context/BombayLoginContext';
import { useModelCollection } from '../../../../../Hooks/useModelCollection';

export function PickerList({ pickModel, isOpen, initialCollection }) {
  const topRef = createRef();

  const loginState = useContext(BombayLoginContext);

  const [isMounted, setIsMounted] = useState(false);
  const [listCollection] = useState(initialCollection);
  const { refreshCollection } = useModelCollection({
    initialCollection: listCollection,
    topRef,
    isMounted,
  });

  useEffect(() => {
    setIsMounted(true);
  }, []);

  function clickHandler(evt, model) {
    evt.preventDefault();
    pickModel(model);
  }

  if (!isOpen) return null;

  return (
    <div className="picker-component" data-testid="picker-component">
      <div className="list-container" ref={topRef}>
        <ul className="picker-list">
          {!listCollection
            ? ''
            : listCollection.map((model) => {
                const key = `mode-list-${model.id}`;
                return (
                  <li
                    className="picker-item"
                    data-testid="picker-item"
                    key={key}
                    data-model-id={model.id}
                    onClick={(evt) => {
                      clickHandler(evt, model);
                    }}
                  >
                    {model.name}
                  </li>
                );
              })}
        </ul>
      </div>
    </div>
  );
}

PickerList.propTypes = {
  pickModel: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
  collectionClass: PropTypes.func.isRequired,
};

export default PickerList;
