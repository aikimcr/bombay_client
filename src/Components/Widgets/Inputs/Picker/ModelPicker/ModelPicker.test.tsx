import React from 'react';
import { act, fireEvent, render, screen } from '@testing-library/react';
import { http, HttpResponse } from 'msw';
import { setupServer } from 'msw/node';

import {
  ContextChanger,
  TestCollectionOne,
  TestCollectionOneURL,
} from '../../../../../testHelpers';
import { setupTestCollectionOneFetch } from '../../../../../Model/testing';
import { mockIntersectionObserver } from '../../../../../Hooks/testing';

import { ModelPicker } from './ModelPicker';

const fetchItems: Record<
  string,
  ReturnType<typeof setupTestCollectionOneFetch>
> = {};

const handlers = [
  http.get(TestCollectionOneURL, ({ request }) => {
    const requestUrl = new URL(request.url);

    if (requestUrl.searchParams.size === 0) {
      const [_getPromise, _fetchBody, _models, fetchDef] = fetchItems.offset0;
      return HttpResponse.json(fetchDef);
    }

    const offset = requestUrl.searchParams.get('offset');
    const offsetKey = `offset${offset || '0'}`;
    const [_getPromise, _fetchBody, _models, fetchDef] = fetchItems[offsetKey];
    return HttpResponse.json(fetchDef);
  }),
];

const server = setupServer(...handlers);

beforeAll(() => {
  fetchItems.offset0 = setupTestCollectionOneFetch(TestCollectionOneURL);
  fetchItems.offset10 = setupTestCollectionOneFetch({
    TestCollectionOneURL,
    offset: 10,
    limit: 10,
  });
  fetchItems.offset20 = setupTestCollectionOneFetch({
    TestCollectionOneURL,
    offset: 20,
    limit: 20,
  });
});

afterAll(() => server.close());

beforeEach(() => {
  window.IntersectionObserver = mockIntersectionObserver;
  server.listen();
});
afterEach(() => {
  server.resetHandlers();
  delete window.IntersectionObserver;
  mockIntersectionObserver._clear();
});

describe('Test ModelPicker', () => {
  it('should match snapshot', async () => {
    const testCollection = new TestCollectionOne({});
    const { asFragment } = render(
      <ContextChanger initialLoggedIn={true}>
        <ModelPicker id="test-picker" initialCollection={testCollection} />
      </ContextChanger>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should contain a picker button and a model picker list', async () => {
    const testCollection = new TestCollectionOne({});
    render(
      <ContextChanger initialLoggedIn={true}>
        <ModelPicker id="test-picker" initialCollection={testCollection} />
      </ContextChanger>,
    );

    const modelPicker = screen.getByTestId('model-picker');
    expect(modelPicker).toBeInTheDocument();
    expect(modelPicker).toHaveAttribute('id', 'test-picker');

    const button = screen.getByTestId('picker-button');
    expect(button).toBeInTheDocument();
    expect(modelPicker).toContainElement(button);

    const list = screen.getByTestId('model-picker-list');
    expect(list).toBeInTheDocument();
    expect(modelPicker).toContainElement(list);
  });

  it('the picker button and the picker list should have their refence attributes set correctly', async () => {
    const testCollection = new TestCollectionOne({});
    const buttonId = 'test-picker-button';
    const listId = 'test-picker-popover';

    render(
      <ContextChanger initialLoggedIn={true}>
        <ModelPicker id="test-picker" initialCollection={testCollection} />
      </ContextChanger>,
    );

    const button = screen.getByTestId('picker-button');
    expect(button).toHaveAttribute('id', buttonId);
    expect(button).toHaveAttribute('aria-expanded', 'false');
    expect(button).toHaveAttribute('aria-controls', listId);
    expect(button).toHaveAttribute('popovertarget', listId);

    const list = screen.getByTestId('model-picker-list');
    expect(list).toHaveAttribute('id', listId);
    expect(list).not.toHaveAttribute('open');
  });

  it('Toggles open state', async () => {
    const testCollection = new TestCollectionOne({});

    render(
      <ContextChanger initialLoggedIn={true}>
        <ModelPicker id="test-picker" initialCollection={testCollection} />
      </ContextChanger>,
    );

    // Jest dome does not seem to handle popovers well.
    const button = screen.getByTestId('picker-button');
    expect(button).toHaveAttribute('aria-expanded', 'false');

    await act(async () => {
      fireEvent.click(button);
    });

    expect(button).toHaveAttribute('aria-expanded', 'true');

    await act(async () => {
      fireEvent.click(button);
    });

    expect(button).toHaveAttribute('aria-expanded', 'false');
  });
});
