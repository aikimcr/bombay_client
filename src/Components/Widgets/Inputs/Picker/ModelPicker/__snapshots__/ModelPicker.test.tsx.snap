// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Test ModelPicker should match snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="model-picker"
    id="test-picker"
  >
    <button
      aria-controls="test-picker-popover"
      aria-expanded="false"
      aria-haspopup="menu"
      class="picker-button"
      data-testid="picker-button"
      id="test-picker-button"
      popovertarget="test-picker-popover"
      popovertargetaction="toggle"
      type="button"
    >
      dummy
      <i
        class="fa-solid fa-caret-down"
      />
    </button>
    <dialog
      class="model-picker-list"
      data-testid="model-picker-list"
      id="test-picker-popover"
      popover="auto"
    >
      <ul
        class="model-picker-list__list"
        data-testid="model-picker-list-list"
      />
    </dialog>
  </div>
  <button
    class="change-test-login"
    data-testid="change-test-login"
  >
    Change Login
  </button>
</DocumentFragment>
`;
